File: di-ipv-cimit-stub/lib/src/main/java/uk/gov/di/ipv/core/library/service/PendingMitigationService.java
package uk.gov.di.ipv.core.library.service;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.message.StringMapMessage;
import uk.gov.di.ipv.core.library.model.UserMitigationRequest;
import uk.gov.di.ipv.core.library.persistence.DataStore;
import uk.gov.di.ipv.core.library.persistence.items.CimitStubItem;
import uk.gov.di.ipv.core.library.persistence.items.PendingMitigationItem;

import java.util.Comparator;
import java.util.List;

import static uk.gov.di.ipv.core.library.config.ConfigurationVariable.CIMIT_STUB_TTL;
import static uk.gov.di.ipv.core.library.config.EnvironmentVariable.PENDING_MITIGATIONS_TABLE;

public class PendingMitigationService {
    private static final Logger LOGGER = LogManager.getLogger();
    public static final String DESCRIPTION = "description";
    private final DataStore<PendingMitigationItem> dataStore;

    public PendingMitigationService(ConfigService configService) {
        boolean isRunningLocally = configService.isRunningLocally();
        dataStore =
                new DataStore<>(
                        configService.getEnvironmentVariable(PENDING_MITIGATIONS_TABLE),
                        PendingMitigationItem.class,
                        DataStore.getClient(isRunningLocally),
                        isRunningLocally,
                        configService);
    }

    public PendingMitigationService(DataStore<PendingMitigationItem> dataStore) {
        this.dataStore = dataStore;
    }

    public void persistPendingMitigation(
            UserMitigationRequest userMitigationRequest, String ci, String method) {
        LOGGER.info(
                new StringMapMessage()
                        .with(DESCRIPTION, "Creating pending mitigation")
                        .with("request", userMitigationRequest.toString())
                        .with("ci", ci)
                        .with("requestMethod", method));
        dataStore.create(
                PendingMitigationItem.fromMitigationRequestAndMethod(
                        userMitigationRequest, ci, method),
                CIMIT_STUB_TTL);
    }

    public void completePendingMitigation(
            String jwtId, String userId, CimitStubItemService cimitService) {
        PendingMitigationItem pendingMitigationItem = dataStore.getItem(jwtId, false);
        if (pendingMitigationItem == null) {
            LOGGER.info(
                    new StringMapMessage()
                            .with(DESCRIPTION, "No pending mitigations found")
                            .with("jwtId", jwtId)
                            .with("userId", userId));
            return;
        }
        List<CimitStubItem> cimitItems =
                cimitService.getCiForUserId(userId, pendingMitigationItem.getMitigatedCi());
        if (cimitItems == null || cimitItems.isEmpty()) {
            LOGGER.warn(
                    new StringMapMessage()
                            .with(DESCRIPTION, "No CI found for attempted mitigation")
                            .with("jwtId", jwtId)
                            .with("userId", userId)
                            .with("ci", pendingMitigationItem.getMitigatedCi()));
            return;
        }

        CimitStubItem itemToMitigate =
                cimitItems.stream()
                        .sorted(Comparator.comparing(CimitStubItem::getIssuanceDate))
                        .reduce((first, second) -> second)
                        .orElseThrow();

        switch (pendingMitigationItem.getRequestMethod()) {
            case "PUT" -> itemToMitigate.setMitigations(pendingMitigationItem.getMitigationCodes());
            case "POST" -> itemToMitigate.addMitigations(
                    pendingMitigationItem.getMitigationCodes());
            default -> throw new IllegalArgumentException(
                    String.format(
                            "Method not supported: %s", pendingMitigationItem.getRequestMethod()));
        }

        cimitService.updateCimitStubItem(itemToMitigate);
        LOGGER.info(
                new StringMapMessage()
                        .with(DESCRIPTION, "CI mitigated")
                        .with("jwtId", jwtId)
                        .with("userId", userId)
                        .with("ci", pendingMitigationItem.getMitigatedCi()));
    }
}

File: di-ipv-cimit-stub-ts/lambdas/src/common/pendingMitigationService.ts
import { dynamoDBClient } from "../clients/dynamoDBClient";
import * as config from "./configService";
import { marshall, unmarshall } from "@aws-sdk/util-dynamodb";
import * as cimitStubItemService from "./cimitStubItemService";
import { CimitStubItem } from "./contraIndicatorTypes";

interface UserMitigationRequest {
  mitigations: string[];
  vcJti: string;
}

interface PendingMitigationItem {
  vcJti: string;
  mitigatedCi: string;
  mitigationCodes: string[];
  requestMethod: string;
  ttl: number;
}

export async function persistPendingMitigation(
  userMitigationRequest: UserMitigationRequest,
  ci: string,
  method: string,
): Promise<void> {
  console.log("Creating pending mitigation", {
    request: userMitigationRequest,
    ci,
    requestMethod: method,
  });

  const item = fromMitigationRequestAndMethod(
    userMitigationRequest,
    ci,
    method,
  );
  await createPendingMitigationItem(item);
}

export async function completePendingMitigation(
  jwtId: string,
  userId: string,
): Promise<void> {
  const pendingMitigationItem = await getPendingMitigationItem(jwtId);

  if (!pendingMitigationItem) {
    console.log("No pending mitigations found", { jwtId, userId });
    return;
  }

  const cimitItems = await cimitStubItemService.getCiForUserId(
    userId,
    pendingMitigationItem.mitigatedCi,
  );

  if (!cimitItems || cimitItems.length === 0) {
    console.warn("No CI found for attempted mitigation", {
      jwtId,
      userId,
      ci: pendingMitigationItem.mitigatedCi,
    });
    return;
  }

  const itemToMitigate = cimitItems
    .sort((a, b) => a.issuanceDate.localeCompare(b.issuanceDate))
    .pop();

  if (!itemToMitigate) {
    throw new Error("No CI found to mitigate");
  }

  switch (pendingMitigationItem.requestMethod) {
    case "PUT":
      itemToMitigate.mitigations = pendingMitigationItem.mitigationCodes;
      break;
    case "POST":
      addMitigations(itemToMitigate, pendingMitigationItem.mitigationCodes);
      break;
    default:
      throw new Error(
        `Method not supported: ${pendingMitigationItem.requestMethod}`,
      );
  }

  await cimitStubItemService.persistCimitStubItem(itemToMitigate);
  console.log("CI mitigated", {
    jwtId,
    userId,
    ci: pendingMitigationItem.mitigatedCi,
  });
}

function fromMitigationRequestAndMethod(
  request: UserMitigationRequest,
  ci: string,
  method: string,
): PendingMitigationItem {
  return {
    vcJti: request.vcJti,
    mitigatedCi: ci,
    mitigationCodes: request.mitigations,
    requestMethod: method,
    ttl: 0,
  };
}

function addMitigations(item: CimitStubItem, newMitigations: string[]): void {
  if (!item.mitigations) {
    item.mitigations = newMitigations;
    return;
  }
  if (!newMitigations) {
    return;
  }
  // Don't add duplicate mitigations
  item.mitigations = [
    ...new Set([...item.mitigations, ...newMitigations]),
  ].sort();
}

async function createPendingMitigationItem(
  item: PendingMitigationItem,
): Promise<void> {
  const nowInSeconds = Math.floor(Date.now() / 1000);
  const ttlSeconds = await config.getCimitStubTtl();
  item.ttl = nowInSeconds + ttlSeconds;

  await dynamoDBClient.putItem({
    TableName: config.getPendingMitigationsTableName(),
    Item: marshall(item, { removeUndefinedValues: true }),
  });
}

async function getPendingMitigationItem(
  vcJti: string,
): Promise<PendingMitigationItem | null> {
  const result = await dynamoDBClient.getItem({
    TableName: config.getPendingMitigationsTableName(),
    Key: marshall({ vcJti }),
  });

  return result.Item
    ? (unmarshall(result.Item) as PendingMitigationItem)
    : null;
}
